### DISCUSSION TITLE
Integrating Diligent graphics engine

### DISCUSSION THREAD

**Discussion by cosator** [u3d-community/U3D#42] [General]
A few weeks ago I started to add Diligent engine to Uhro3D v1.8 as an alternative graphics backend. Diligent brings Directx 12 and Vulkan support on top of Directx 11 and OpenGL, one shader language for all APIs, support for compute, geometry, hull and other shader types (for future expansion of U3D/Urho3D capabilities).
I have most of the samples working, but there is some work ahead to get all the samples working properly for all APIs. Also some optimizations will be needed once all the features are implemented.
The original Directx 11 and OpenGL graphics are still working without modification. Also, I did not modify anything on the rendering side.
Is there any interest here to add Diligent as an option for graphics backend?

**Comment by SirNate0** [comment]
> Directx 12 and Vulkan support ..., one shader language for all APIs, support for compute, geometry, hull and other shader types

All of these sound pretty good to me, particularly the other shader types. Though checking briefly, I see a lot of issues in the support for converting HLSL to GLSL - are these common in practice? https://diligentgraphics.com/diligent-engine/shader-converter/

**Reply by cosator** [reply]
It's true that there are some challenges for converting HLSL to GLSL for OpenGL support. For Vulkan it's much easier because it is using Spir-V and there are almost no issues. For HLSL to GLSL conversion, I might need to do some preprocessing of HLSL shaders before conversion, but I don't thing it's going to be extremely challenging or resource intensive.

**Comment by rbnpontes** [comment]
My two cents here. I had worked at initial stages on diligent engine integration into rbfx and i ran into some issues through the development that you guys should consider. 

First,  HLSL to GLSL tool from Diligent will not work with current urho HLSL shaders. basically you can't enable and disable vertex inputs through macros on Vulkan and OpenGL backends, you need to do a lot of workarounds to make fully work.
I recommend to work only with GLSL and you can use glslang to generate an spirv code and from spirv code you can get a HLSL code with Spirv Cross.
Another issue you will deal is lack of reflections on diligent engine. You must need reflect shader vertex attributes to make vertex buffer work. But if you choose to deal with glslang -> spirv you can easy reflect these parameters. or you can modify diligent code to expose these parameters like Eugene does on rbfx.

And the final part, you must generate every state for each combination of urho graphics parameters. 
So i made this:  
1. For each parameter i generate a pipeline hash than i lookup into a pipeline hash table
2. For each texture unit assign on graphics, i generate a second hash with combination of pipeline hash to get proper Shader Resource Binding. (If they don't exist, then i create a new one).

Urho3D renderer is obsolete but its possible to integrate diligent engine with no problems. But you should consider to rewrite the whole render like Eugene does on rbfx.
I started to do this same work again on my Atomic Engine fork: https://github.com/rbnpontes/REngine-Atomic
You can check at `feat/diligent branch`

**Reply by cosator** [reply]
I solved all the problems for DX11m DX12 and Vulkan (using HLSL for shaders). I used U3D fork of Urho3D, but it is almost identical with Urho3D. I got part of OpenGL working, but unfortunately I did not have much time lately to fix the remaining issues.
If anybody is interested, this is the repo: https://github.com/cosator/U3D_Diligent

**Comment by ZirPlatinumss** [comment]
Ha, I literally had no idea this existed! I independently started integrating DiligentEngine into U3D today and got DX12 working with 3D rendering (textured geometry, normal maps, per-pixel lighting, instancing, UI) in a single day. Funny how I ran into the same problems and came up with similar solutions — PSO caching, SRB management, constant buffer system.

@cosator would love to compare approaches. I'll check out your repo. Great to know someone else saw the same path forward for U3D.

**Reply by cosator** [reply]
I was planning to finish OpenGL support this year. All the samples work with Vulkan, DX11, and DX12. OpenGL had some issues with the shadow maps, but with AI help that should not be too challenging.
